src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(chip::Messaging::ExchangeContext * ec,
+                                                       const chip::PayloadHeader & payloadHeader,
+                                                       chip::System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    ChipLogDetail(BDX, " %s: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId, __FUNCTION__,
+                  payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+
+    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
```

Comment:
OK, but do we actually know which event types signify the end of traffic on the exchange?  It sure would be nice if the exchange management were _not_ forced on our consumers.

Reply:
This was existing code copied from the old ota provider delegate bridge. I didn't check the comment. Happy to file an issue and work on this as a follow up if that works for you. let me know.

Reply:
This is not applicable anymore since we are closing the exchange when transfer ends or times out or an error occurs during BDX